Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1375 @@
+
+```hs
+data Error i e
+  = EndOfInput  -- Expected more input, but there is nothing
+  | Unexpected i  -- We didn't expect to find this element
+  | CustomError e  -- Extra errors the user may want to create
+  | Empty  -- Used in `Alternative` implementation of `empty`
+  deriving (Eq, Show)
+
+newtype Parser i e a = Parser
+  { runParser :: [i] -> Either [Error i e] (a, [i])
+  }
+```
+
+This is the meaning of each of the type variables:
+* `i`: The input stream. For most cases, we will use `Char`, as `[Char]` is the same as `String`. It's interesting to notice that keeping this as a type variable allows us to parse things that are not strings as well.
+* `e`: The type of custom error messages. If we don't have those, we may use `Void` instead.
+* `a`: The result of our parsing function. It represents the structure parsed from our consumed input.
+
+Let's begin by creating the most primitive function: `satisfy`. It takes some predicate that operates on the current character, and advances the parser, if possible, returning the consumed character.
+
+We must take care that we are not at the end of the input, however, in which case we should fail.
+
+```hs
+satisfy :: (i -> Bool) -> Parser i e i
+satisfy predicate = Parser $ \input ->
+  case input of
+    [] -> Left [EndOfInput]
+    hd : rest
+      | predicate hd -> Right (hd, rest)
+      | otherwise    -> Left [Unexpected hd]
+```
+
+It's also common to define a function that only succeeds if the input strictly matches some given element.
+
+```hs
+char :: Eq i => i -> Parser i e i
+char i = satisfy (== i)
+```
+
+If you load this in GHCi, you should now be able to test it:
+```hs
+>>> :l Parser
+
+>>> runParser (char 'h') "hello"
+Right ('h', "ello")
+
+>>> runParser (char 'h') "greetings"
+Left [Unexpected 'g']
+
+>>> runParser (char 'h') ""
+Left [EndOfInput]
+```
+
+Let's now create `Monad` and `Alternative` instances for our parser. For educative purposes, we will also explicitly implement `Functor` and `Applicative` instead of `fmap = liftM` and `(<*>) = ap`.
+
+We start with `Functor`. Since we map over the `a` type variable obtained in the case of a successful parse, we can only apply `f` in case if we get `Right`.
+
+```hs
+instance Functor (Parser i e) where
+  fmap f (Parser p) = Parser $ \input ->
+    case p input of
+      Left err -> Left err
+      Right (output, rest) -> Right (f output, rest)
+```
+
+Here's a little shorter implementation. Since `Either` is a monad, we can use do-notation.
+
+```hs
+instance Functor (Parser i e) where
+  fmap f (Parser p) = Parser $ \input -> do
+    (output, rest) <- p input
+    pure (f output, rest)
+```
+
+It's interesting to notice that `Either` is a monad, so we can also write the definition of most things more concisely, as shown in the commented part.
```

Comment:
You mentioned that in the previous line already, sounds a bit awkward now. :D

Reply:
Some leftover, I removed this part.